[doc] Added onboarding doc for what a new project needs to do to to g… (#60)
godaddy/warehouse.ai: onboaring.md

1. [Update your `.npmrc` configuration](#2-update-npmrc-configuration)
1. [Configure your module for Warehouse](#3-configure-your-module-for-warehouse)
    1. [Add properties to package.json](#add-properties-to-packagejson)
    1. [Set publish target](#warehouse-as-publish-target)
    1. [Setup `wrhs.toml`](#setup-wrhstoml)
1. [Publishing your module](#4-publishing-your-module)
    1. [Monitoring build progress](#monitoring-build-progress)
    1. [Promoting your module](#promote-your-module-to-testprod)
1. [Using build output](#5-using-build-output)
    1. [Inspect your bundle](#inspect-your-bundle)
    1. [Serving assets from a web service](#serving-assets-from-a-web-service)
1. [Advanced features](#6-advanced-features)
    - [Building for multiple locales](#building-for-multiple-locales)
    - [Debugging a build](#debugging-a-build)
    - [Minification](#minification)
    - [Dependent builds](#dependent-builds)

### 1. Get credentials

Warehouse is hosted on AWS. Performing requests against the API requires authentication. You will need to add the credentials, which you created as part of the setup process for your running instance of `warehouse.ai`, to your `.npmrc` file. 

> ### 2. Update npmrc configuration

Reviewer: Agreed. 1 & 2 are really "Configure your `npm` client". The flow is solid tho.

`output[.min].css` respectively.

```ini
[files]
dev = ['dist/ouput.js', 'dist/output.css']
test = ['dist/ouput.js', 'dist/output.css']
prod = ['dist/ouput.min.js', 'dist/output.min.css']
```
> We plan to do all configuration from this file in the future.
> This will allow us to expose more features and make setup easier.

From this example you can discern that uploaded files are configurable per environment and file paths are
determined relative to the root of your module. The `dist` directory is not an arbitrary choice, `warehouse.ai` will need the `dist` directory to explicitly exist so it knows which files to serve.

Finally, if you are using `webpack`, you will need a `webpack.config.js` in the root directory. Make sure that files in your `wrhs.toml` match the ouput of your `webpack.config.js`. It is important to note that all `warehouse.ai` is doing is to call `webpack` in this case, thus all of your configuration must live within this file (not as command line arguments). 

```js
const path = require('path');

module.exports = {
  entry: './src/index.js',
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'output.js'
  }
};
```

> #### Extract Config

Reviewer 2: I'd just delete this section, as part of the work to support the new module for simplified config, we should come back and update these docs.

COMING SOON: You will be able to extract your warehouse.ai configuration to keep it in one place by using [extract-config]. 

> ### 4. Publishing your module

Reviewer 2: I feel like at this point (at the end of the last step), this document is done and can just point to other documents. something like

```md
That's it, your package is setup to use `warehouse.ai`.

You can now publish your package with `npm publish` as you normally would and use npm dist-tags to promote your build through environments. See [common commands][new doc here] for help with how to do that.

There are also more [advanced configuration options][another new doc] that allow you to control minification, create dependent builds, create locale-specific builds, etc.

You can also use [`warehouse.ai-api-client`] in your application to consume the [build output][more new doc here].
```

The new docs can just be the content that you have below, just organized into smaller, more targeted docs.

Author: oops, missed this comment in the first round, making requested changes for it now.

Publish your package as you normally would:

```sh-session
npm version <newversion> 

npm publish
```

This will trigger a build in Warehouse and publish the module to the `write` endpoint you specified. Initially the build of this version
will only be available in `DEV`. Subsequently the files are only uploaded to the `DEV` CDN as well. A hash is generated
from the file content that the browser will use to do caching based on file content. Warehouse provides
various mechanics to retrieve the build files. We'll [dive into the mechanisms
later](#using-build-output).

#### Monitoring build progress

To check the status of the build process you can either use the [Warehouse UI][ui] or
[CLI](cli).
Once you have `Warehouse UI` set up, it will show the
build progress for each environment. Wait for the `DEV` build to complete.

#### Promote your module to TEST/PROD

After the initial publish, the build will only be available in the `DEV` environment. Assuming the bundle
works as expected, it can be promoted to `TEST` and/or `PROD`. Promoting your module's bundle to a specific
environment can be done through `npm dist-tag`. For example, to promote your bundle to `TEST` run:

```

> npm dist-tag add [module-name]@[semver] test --reg=https://wherever-you-deployed-warehouse.ai

Reviewer 2: ```suggestion
npm dist-tag add [module-name]@[semver] test --registry=https://wherever-you-deployed-warehouse.ai
```

We should spell out all the commands in docs, even when short forms exist. That way anyone not familiar with the short forms can still understand the docs and what the arguments are for.

```

Where `[module-name]` is the package name of your module and `[semver]` the exact semver that was published to
`DEV`. Also be aware that the `--reg` argument is required, since `publishConfig` only applies to `npm publish`.
To promote your module to `PROD`, simply `npm dist-tag` again but replace `test` with `prod`.

### 5. Using build output

Now that you can promote your module's bundle to each environment, you can start using the bundle.
First, let's inspect the available bundle output. Then, we'll reference the assets from your web application.

#### Inspect your bundle

> [TODO] The easiest way to quickly inspect a build and files on CDN is by using the Warehouse CLI. Install the CLI by

Reviewer 2: Is this still `TODO`?

following [the directions from the readme][cli].

Now let's inspect your build by running the following command. Replace `[module-name]` with your package name.
This will return bundle and build details for your package in `dev`. Similarly you can inspect your bundle in
`test` and `prod`.

```sh-session
wrhs get:build [module-name] dev
```

#### Serving assets from a web service

Use the [Warehouse-client][client] to retrieve bundle information for your webserver's responses. The client
requires the same authentication and endpoint details as provided to the [CLI](#inspect-your-bundle).
Currently, we only have a client available for Node.JS. Usage is straightforward. Install the module as

> a `dependency` of your webserver, e.g. `npm i --save warehouse.ai-api-client`.

Reviewer 2: ```suggestion
a `dependency` of your webserver, e.g. `npm install warehouse.ai-api-client`.
```

> Next, setup an instance of the client like below. Ideally the `[username]` and `[password]` are stored in a secure config, for example by using

Reviewer 2: we should probably reword this and fixup the example below a bit since the auth is optional based on configuration of the `warehouse.ai` instance.

  retry: {
    max: 3000
  }
});
```

This client can now be used to get information from bundle `details`. This object will contain properties like
`files` and `fingerprints` (unique hashes). Which enables you to reference the correct CDN assets in your
webserver responses.

```js
warehouse.builds.get({
  env: 'prod',                       // This should match the enviroment your server is running in
  pkg: '[module-name]',              // Replace this with your package's name
  locale: 'en-US'                    // Defaults to `en-US`, but could be any locale you specified as target
}, function done(error, details) {
  console.log(details);              // Will return properties similar to the CLI.
});
```

To make sure that build information is retrieved out-of-band, the client has build in memory and/or file caching
mechanisms. Bundles information will be mostly static between requests. The client will automatically
perform update requests on an interval against Warehouse to retrieve the bundle's latests details.

### 6. Advanced features

#### Building for multiple locales

Warehouse can create bundles per locale and/or

> market. This should result in optimized bundles per locale. By default any package is built for `en-US`. To enable builds for multiple locales, add the `locales` property

Reviewer 2: can you clean up the line breaks while you're here? :)

of your module's `package.json`. `locales` should be specified as an `array`. For example:

```js
"locales": [
  "da",
  "de",
  "nl-NL",
  ...
],
```

Locales can be specified with their full [IETF language tag][ietf], which is composed of a [primary language subtag][language-subtag] and a [region subtag][region-subtag], or their primary language subtag alone.

Each locale will run as a separate build. The build will set the locale to environment variables. These are
`process.env.WRHS_LOCALE` and `process.env.LOCALE`, preferably use the latter. You can refer to
the environment variables directly in your webpack configuration.

#### Debugging a build

If your build failed to complete for some reason, you can first inspect the result of each step
through the build status. Run the following command to get a detailed description of each step.

```js

> wrhs get:status @[moduleName] dev -e

Reviewer 2: ```suggestion
wrhs get:status [moduleName] dev -e
```

the `@` is part of the module name.

```

> The results from a successful build can be found below. If one of the steps fails. Warehouse will retry the

Reviewer 2: ```suggestion
The results from a successful build can be found below. Warehouse has some built-in retries for some steps.
```

> build step, as can be seen from the `ERROR` on install. 

Reviewer 2: ```suggestion
This can be seen in the `wrhs` CLI output below with the `ERROR` for  `npm install`. 
```
